### PR TITLE
Pack topics

### PR DESCRIPTION
**Pull request by aaronbuchwald**
## Why this should be merged

This PR adds `PackTopics` to the abi package to support packing events. This is intended to be made available for precompiles to pack events, so that they can create logs.

## How this works

This works by using the existing code for packing topics used in the abi package to support creating filter queries and breaks out the actual creation of a topic hash into a helper function, so that we can re-purpose it to create an array of topics as would be created by an event log rather than the type `[][]common.Hash` that would be used when creating a filter query.

## How this was tested

Adds a unit test in `abi_test.go` to test the encoding of an event without indexed args, with indexed args, and of an anonymous event.

## Reference for Event Encoding

https://docs.soliditylang.org/en/v0.8.17/abi-spec.html#indexed-event-encoding

### CODE REVIEW

File: accounts/abi/topics.go [hunk 1]
```diff
@@ -37,71 +37,99 @@ import (
...
+			return common.Hash{}, fmt.Errorf("unsupported indexed type: %T", rule)
+		}
+	}
+	return topic, nil
+}
+
+// PackTopics attempts to pack the array of filters into an array of corresponding topics
+// according to the Solidity documentation.
+// Note: PackTopics does not support array(both fixed size and dynamic-size) and struct types.
+func PackTopics(filter []interface{}) ([]common.Hash, error) {
```
Comment (aaronbuchwald) [inline accounts/abi/topics.go:110]: We should move this into the ABI package and have a single function to pack both the topics and event input data.

File: accounts/abi/topics.go [hunk 2]
```diff
@@ -37,71 +37,99 @@ import (
...
-				case val.Kind() == reflect.Array && reflect.TypeOf(rule).Elem().Kind() == reflect.Uint8:
-					reflect.Copy(reflect.ValueOf(topic[:val.Len()]), val)
-				default:
-					return nil, fmt.Errorf("unsupported indexed type: %T", rule)
-				}
+			topic, err := packTopic(rule)
+			if err != nil {
+				return nil, err
 			}
+
```
Comment (darioush) [inline accounts/abi/topics.go:132]: nit: could we remove the wspace to minimize the diff?

File: accounts/abi/topics.go [hunk 3]
```diff
@@ -37,71 +37,99 @@ import (
...
+		case val.Kind() == reflect.Array && reflect.TypeOf(rule).Elem().Kind() == reflect.Uint8:
+			reflect.Copy(reflect.ValueOf(topic[:val.Len()]), val)
+		default:
+			return common.Hash{}, fmt.Errorf("unsupported indexed type: %T", rule)
+		}
+	}
+	return topic, nil
+}
+
+// PackTopics attempts to pack the array of filters into an array of corresponding topics
```
Comment (darioush) [inline accounts/abi/topics.go:107]:
```suggestion
// PackTopics packs the array of filters into an array of corresponding topics
```

File: accounts/abi/topics.go [hunk 4]
```diff
@@ -37,71 +37,99 @@ import (
...
+// according to the Solidity documentation.
+// Note: PackTopics does not support array(both fixed size and dynamic-size) and struct types.
```
Comment (darioush) [inline accounts/abi/topics.go:109]:
```suggestion
// Note: PackTopics does not support array (fixed or dynamic-size) or struct types.
```

File: accounts/abi/topics.go (hunk 1 above)
Comment (darioush) [reply accounts/abi/topics.go:110]: not sure I follow this comment.

Comment (aaronbuchwald) [reply accounts/abi/topics.go:110]: Updated the PR with what I was thinking.

Logs include both topics `[]common.Hash` and event data, which is more similar to the way function arguments are encoded and `PackTopics` only converts arguments to the corresponding topic hash based off of this.

Now there is `PackEvent` which will take in all of the arguments both indexed and non-indexed and return both the resulting topics and event data.

File: accounts/abi/abi.go [hunk 5]
```diff
@@ -91,6 +91,54 @@ func (abi ABI) Pack(name string, args ...interface{}) ([]byte, error) {
 	return append(method.ID, arguments...), nil
 }
 
+// PackEvent packs the given event name to conform the ABI.
```
Comment (darioush) [inline accounts/abi/abi.go:94]:
```suggestion
// PackEvent packs the given event name and arguments to conform the ABI.
```

File: accounts/abi/abi_test.go [hunk 6]
```diff
@@ -39,6 +39,7 @@ import (
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/math"
 	"github.com/ethereum/go-ethereum/crypto"
+	"gotest.tools/assert"
```
Comment (darioush) [inline accounts/abi/abi_test.go:42]: Usually we import stretchr

File: accounts/abi/topics.go [hunk 7]
```diff
@@ -37,70 +37,97 @@ import (
 	"github.com/ethereum/go-ethereum/crypto"
 )
 
+// packTopic attempts to pack rule into the corresponding hash value for a log's topic
```
Comment (darioush) [inline accounts/abi/topics.go:40]:
```suggestion
// packTopic packs rule into the corresponding hash value for a log's topic
```

**Review by darioush — APPROVED** [review]
LGTM after the test import is fixed, other comments optional

File: accounts/abi/abi_test.go (hunk 6 above)
Comment (aaronbuchwald) [reply accounts/abi/abi_test.go:42]: Nice catch

**Comment by aaronbuchwald** [comment]
Added one more simplifying change to this PR to have only one argument array passed in to `PackEvent`, so I re-requested your guys' reviews @ceyonur @darioush

**Review by darioush — APPROVED** [review]

**Pull request merged**